backend/src/controller/user-controller.ts

```diff
@@ -0,0 +1,215 @@
+import { Request, Response } from "express";
+import { z } from "zod";
+import prisma from "../prisma-client";
+
+const UserRequestSchema = z.object({
+  email: z.string().email().nonempty(),
+  username: z.string().nonempty(),
+  firstName: z.string().nonempty(),
+  lastName: z.string().nonempty(),
+});
+
+type UserDto = z.infer<typeof UserRequestSchema>;
+
+const Errors = {
+  UsernameAlreadyTaken: "UserNameAlreadyTaken",
+  EmailAlreadyInUse: "EmailAlreadyInUse",
+  ValidationError: "ValidationError",
+  ServerError: "ServerError",
+  ClientError: "ClientError",
+  UserNotFound: "UserNotFound",
+} as const; // 'as const' ensures these values are treated as string literals
+
+type ErrorType = (typeof Errors)[keyof typeof Errors];
```

Comment:
It could be an enum instead:

```ts
enum Errors {
  UsernameAlreadyTaken = "UserNameAlreadyTaken",
  EmailAlreadyInUse = "EmailAlreadyInUse",
  ValidationError = "ValidationError",
  ServerError = "ServerError",
  ClientError = "ClientError",
  UserNotFound = "UserNotFound",
}
```

Then you don't need `type ErrorType = (typeof Errors)[keyof typeof Errors];` at all.